eth/downloader/downloader_test.go

```diff
@@ -1541,3 +1530,42 @@ func testFakedSyncProgress(t *testing.T, protocol int, mode SyncMode) {
 		t.Fatalf("Final progress mismatch: have %v/%v/%v, want 0-%v/%v/%v", origin, current, latest, targetBlocks, targetBlocks, targetBlocks)
 	}
 }
+
+// This test reproduces an issue where unexpected deliveries would
+// block indefinitely if they arrived at the right time.
+func TestDeliverHeadersHang62(t *testing.T) {
```

Comment:
I don't know how to do that without writing 6 almost identical tests.
The specific part that is being tested is the interaction between `spawnSync` and `deliver`.
These functions are generic and do not care about the sync mode.